### PR TITLE
[Doc] Add Distributed Testing Example for `pl.Trainer.test()`

### LINKED ISSUE #35225: [Train] LightningTrainer supports `pl.Trainer.test()` for batch inference example
### Description

Currently, the users have to set aside a dedicated GPU solely for the `pl.Trainer.test` method because the `LightningTrainer` uses autoscaling for the training process, while has no support for inference. 

- The users still want to leverage the native pytorch lightning interfaces like `on_test_batch()`, `on_test_epoch_end()`.
- It's not convenient for them to create a batch predictor to do batch inference if they are not highly relied on Ray Data.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

Currently LightningTrainer only supports distributed training. However, some users complained that they don't know how to use `pl.Trainer.test()` for testing in a distributed manner. 

- We are not planning to add a `LightningTrainer.test()` API for users, since it will introduce inconsistency among other trainers and undefined behavior for auto-restoration.
- This PR adds an example that demonstrate how to use `init_torch_dist_process_group` + `TorchDistributedWorker` for distributed testing.

## Related issue number

Closes #35225
## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 1]
```diff
@@ -1,7 +1,6 @@
 {
  "cells": [
   {
-   "attachments": {},
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:4]: Remove `datamodule` import at top

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 2]
```diff
@@ -23,9 +22,18 @@
...
-   "execution_count": null,
+   "execution_count": 2,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/ray/anaconda3/lib/python3.9/site-packages/xgboost/compat.py:31: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:32]: Remove this output

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 3]
```diff
@@ -603,7 +878,15 @@
    "metadata": {
     "tags": []
    },
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Global seed set to 888\n"
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:886]: Probably don't need to include this line

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 4]
```diff
@@ -612,6 +895,15 @@
   {
    "cell_type": "code",
    "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "datamodule.setup()"
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:901]: Merge this with the following block.

```python
datamodule.setup() # explain why you need this
test_dataloader = datamodule.test_dataloader()

trainer = pl.Trainer()
result = trainer.test(best_model, dataloaders=test_dataloader)

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 5]
```diff
@@ -686,7 +977,157 @@
    ]
   },
   {
-   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can create a set of Ray Actors for distributed inference using multiple GPUs. Here we demonstrate how to perform distributed testing with 4 GPUs."
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:983]: Add a new header before this section to make it clear that it is a different task.

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 6]
```diff
@@ -763,7 +1210,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.13"
+   "version": "3.9.15"
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1213]: Remove LightningPredictor for now since we have the other options.

**Comment by woshiyyya** [comment]
Addressed the comments.

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 7]
```diff
@@ -690,51 +970,158 @@
...
+    "import ray\n",
+    "import pytorch_lightning as pl\n",
     "\n",
-    "predictor = LightningPredictor.from_checkpoint(\n",
-    "    checkpoint, MNISTClassifier, use_gpu=use_gpu\n",
+    "from pytorch_lightning.plugins.environments.lightning_environment import (\n",
+    "    LightningEnvironment,\n",
     ")\n",
+    "from ray.air.util.torch_dist import (\n",
+    "    TorchDistributedWorker,\n",
```
Comment (richardliaw) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1065]: is this a public API now @matthewdeng ?

Comment (matthewdeng) [reply doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1065]: It is not at the moment. Nothing has changed since it was [introduced](https://github.com/ray-project/ray/pull/34202).

Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1065]: Actually, we are already using it in the [OPT-30B Batch Inference](https://docs.ray.io/en/master/ray-air/examples/opt_deepspeed_batch_inference.html?highlight=OPT%20#define-a-deepspeed-predictor) example.

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 8]
```diff
@@ -690,51 +969,158 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "You can also use {class}`LightningPredictor <ray.train.lightning.LightningPredictor>` for inference."
+    "## Multi-nodes Testing\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:972]:
```suggestion
    "## Multi-node Testing\n",
```

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 9]
```diff
@@ -690,51 +969,158 @@
...
+    "\n",
+    "Alternatively, if you have a large test set and want to speed up the testing process in parallel, you can create a group of Ray Actors to leverage multi-nodes and multi-GPUs for distributed inference. Here we demonstrate how to set up a process group and do evaluation using 4 GPUs."
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:974]:
```suggestion
    "Alternatively, if you have a large test set and want to speed up the testing process in parallel, you can create a group of Ray Actors to leverage multiple GPUs across multiple nodes for distributed inference. Here we demonstrate how to set up a process group and do evaluation using 4 GPUs."
```

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 10]
```diff
@@ -690,51 +969,158 @@
...
+    "@ray.remote\n",
+    "class TestWorker(TorchDistributedWorker):\n",
+    "    def run(self):\n",
+    "        trainer = pl.Trainer(\n",
+    "            num_nodes=num_workers,\n",
+    "            accelerator=\"gpu\",\n",
+    "            strategy=\"ddp\",\n",
+    "            plugins=[RayEnvironment()],\n",
+    "        )\n",
+    "        return trainer.test(best_model, datamodule=datamodule)\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1105]: For consistency with the single-node example above, should we just pass in `dataloaders` instead?

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 11]
```diff
@@ -690,51 +969,158 @@
...
+    "    LightningEnvironment,\n",
     ")\n",
+    "from ray.air.util.torch_dist import (\n",
+    "    TorchDistributedWorker,\n",
+    "    init_torch_dist_process_group,\n",
+    "    shutdown_torch_dist_process_group,\n",
+    ")\n",
+    "\n",
+    "\n",
+    "class RayEnvironment(LightningEnvironment):\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1070]: Note: May be worth considering promoting this to be a public API later

**Review by matthewdeng — APPROVED** [review]

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb (hunk 11 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1070]: Oh actually this is not the same `RayEnvironment` we are using for LightningTrainer. We are using env variable to determine local/global ranks, but that one uses `session.get_local_rank()`

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb (hunk 10 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_mnist_example.ipynb:1105]: Make sense. I'll change it

**Review by matthewdeng — APPROVED** [review]
